README.md
pfcd
====

[![Build Status](https://travis-ci.org/picfight/pfcd.png?branch=master)](https://travis-ci.org/picfight/pfcd)
[![ISC License](http://img.shields.io/badge/license-ISC-blue.svg)](http://copyfree.org)
[![GoDoc](https://img.shields.io/badge/godoc-reference-blue.svg)](http://godoc.org/github.com/picfight/pfcd)
[![Go Report Card](https://goreportcard.com/badge/github.com/picfight/pfcd)](https://goreportcard.com/report/github.com/picfight/pfcd)

## PicFight coin overview

PicFight coin is a [Decred](https://decred.org)-based cryptocurrency.
It utilizes a hybrid proof-of-work and proof-of-stake mining system.
A unit of the currency is called a `picfight coin` (PFC).

https://picfight.org

## What is pfcd?

pfcd is a full node implementation of PicFight coin written in Go (golang).

It acts as a fully-validating chain daemon.  pfcd maintains the entire past
transactional ledger of PicFight coin and allows relaying of transactions
to other PicFight coin nodes around the world.

## What is a full node?

The term 'full node' is short for 'fully-validating node' and refers to software
that fully validates all transactions and blocks, as opposed to trusting a 3rd
party. In addition to validating transactions and blocks, nearly all full nodes
also participate in relaying transactions and blocks to other full nodes around
the world, thus forming the peer-to-peer network.

## Getting Started

So, you've decided to help the network by running a full node.  Great!  Running
pfcd is simple.  All you need to do is install pfcd on a machine that is
connected to the internet and launch it.

Also, make sure your firewall is configured to allow inbound connections to port
9108.

## Installing and updating

### Build from source (all platforms)

Building or updating from source requires the following build dependencies:

- **Go 1.13**

  Installation instructions can be found here: https://golang.org/doc/install.
  It is recommended to add `$GOPATH/bin` to your `PATH` at this point.

- **Git**

  Installation instructions can be found at https://git-scm.com or
  https://gitforwindows.org.

To build and install from a checked-out repo, run `go install . ./cmd/...` in
the repo's root directory.  Some notes:

* Set the `GO111MODULE=on` environment variable if using Go 1.11 and building
  from within `GOPATH`.

* The `pfcd` executable will be installed to `$GOPATH/bin`.  `GOPATH`
  defaults to `$HOME/go` (or `%USERPROFILE%\go` on Windows) if unset.

### Example of obtaining and building from source on Windows:

```bash
git clone https://github.com/picfight/pfcd ~/src/pfcd
cd ~/src/pfcd
set GO111MODULE=on
go build ./...
go install . .\cmd\...
```

### Running Tests

To run the tests locally:

```bash
set GO111MODULE=on
go build ./...
go clean -testcache
go test ./...
```

## Example run commands

Launch mining node:
```bash
pfcd
     --generate
     --miningaddr "JsCVh5SVDQovpW1dswaZNan2mfNWy6uRpPx"
     --listen=127.0.0.1:30000
     --rpclisten=127.0.0.1:30001
     --datadir=nodeA
     --rpccert=nodeA\rpc.cert
     --rpckey=nodeA\rpc.key     
     --txindex
     --addrindex
     --rpcuser=node.user
     --rpcpass=node.pass
```

Launch second node and connect to it the first one for syncing:
```bash
pfcd
     --listen=127.0.0.1:30002
     --rpclisten=127.0.0.1:30003
     --addpeer=127.0.0.1:30000
     --datadir=nodeB
     --rpccert=nodeB\rpc.cert
     --rpckey=nodeB\rpc.key
     --txindex
     --addrindex
     --rpcuser=node.user
     --rpcpass=node.pass
     
```

Enjoy your little blockchain network for a while.

## Contact

If you have any further questions you can find us at the
[integrated github issue tracker](https://github.com/picfight/pfcd/issues).

## Documentation

Since pfcd is a fork of Decred, the documentation for pfcd is located in the
[Decred-docs](https://github.com/decred/dcrd/tree/master/docs) folder.

## License

pfcd is licensed under the [copyfree](http://copyfree.org) ISC License.


CHANGES
============================================================================
User visible changes for pfcd
  A full-node Picfight implementation written in Go
============================================================================

Changes in 0.11.1 (Wed May 27 2015)
  - Protocol and network related changes:
    - Use correct sub-command in reject message for rejected transactions
      (#436, #437)
    - Add a new parameter --torisolation which forces new circuits for each
      connection when using tor (#430)
  - Transaction relay (memory pool) changes:
    - Reduce the default number max number of allowed orphan transactions
      to 1000 (#419)
    - Add a new parameter --maxorphantx which allows the maximum number of
      orphan transactions stored in the mempool to be specified (#419)
  - RPC changes:
    - Modify listtransactions result to include the 'involveswatchonly' and
      'vout' fields (#427)
    - Update getrawtransaction result to omit the 'confirmations' field
      when it is 0 (#420, #422)
    - Update signrawtransaction result to include errors (#423)
  - btcctl utility changes:
    - Add gettxoutproof command (#428)
    - Add verifytxoutproof command (#428)
  - Notable developer-related package changes:
    - The btcec package now provides the ability to perform ECDH
      encryption and decryption (#375)
    - The block and header validation in the blockchain package has been
      split to help pave the way toward concurrent downloads (#386)
  - Misc changes:
    - Minor peer optimization (#433)
  - Contributors (alphabetical order):
    - Dave Collins
    - David Hill
    - Federico Bond
    - Ishbir Singh
    - Josh Rickmar

Changes in 0.11.0 (Wed May 06 2015)
  - Protocol and network related changes:
    - **IMPORTANT: Update is required due to the following point**
    - Correct a few corner cases in script handling which could result in
      forking from the network on non-standard transactions (#425)
    - Add a new checkpoint at block height 352940 (#418)
    - Optimized script execution (#395, #400, #404, #409)
    - Fix a case that could lead stalled syncs (#138, #296)
  - Network address manager changes:
    - Implement eclipse attack countermeasures as proposed in
      http://cs-people.bu.edu/heilman/eclipse (#370, #373)
  - Optional address indexing changes:
    - Fix an issue where a reorg could cause an orderly shutdown when the
      address index is active (#340, #357)
  - Transaction relay (memory pool) changes:
    - Increase maximum allowed space for nulldata transactions to 80 bytes
      (#331)
    - Implement support for the following rules specified by BIP0062:
      - The S value in ECDSA signature must be at most half the curve order
        (rule 5) (#349)
      - Script execution must result in a single non-zero value on the stack
        (rule 6) (#347)
      - NOTE: All 7 rules of BIP0062 are now implemented
    - Use network adjusted time in finalized transaction checks to improve
      consistency across nodes (#332)
    - Process orphan transactions on acceptance of new transactions (#345)
  - RPC changes:
    - Add support for a limited RPC user which is not allowed admin level
      operations on the server (#363)
    - Implement node command for more unified control over connected peers
      (#79, #341)
    - Implement generate command for regtest/simnet to support
      deterministically mining a specified number of blocks (#362, #407)
    - Update searchrawtransactions to return the matching transactions in
      order (#354)
    - Correct an issue with searchrawtransactions where it could return
      duplicates (#346, #354)
    - Increase precision of 'difficulty' field in getblock result to 8
      (#414, #415)
    - Omit 'nextblockhash' field from getblock result when it is empty
      (#416, #417)
    - Add 'id' and 'timeoffset' fields to getpeerinfo result (#335)
  - Websocket changes:
    - Implement new commands stopnotifyspent, stopnotifyreceived,
      stopnotifyblocks, and stopnotifynewtransactions to allow clients to
      cancel notification registrations (#122, #342)
  - btcctl utility changes:
    - A single dash can now be used as an argument to cause that argument to
      be read from stdin (#348)
    - Add generate command
  - Notable developer-related package changes:
    - The new version 2 btcjson package has now replaced the deprecated
      version 1 package (#368)
    - The btcec package now performs all signing using RFC6979 deterministic
      signatures (#358, #360)
    - The txscript package has been significantly cleaned up and had a few
      API changes (#387, #388, #389, #390, #391, #392, #393, #395, #396,
      #400, #403, #404, #405, #406, #408, #409, #410, #412)
    - A new PkScriptLocs function has been added to the wire package MsgTx
      type which provides callers that deal with scripts optimization
      opportunities (#343)
  - Misc changes:
    - Minor wire hashing optimizations (#366, #367)
    - Other minor internal optimizations
  - Contributors (alphabetical order):
    - Alex Akselrod
    - Arne Brutschy
    - Chris Jepson
    - Daniel Krawisz
    - Dave Collins
    - David Hill
    - Jimmy Song
    - Jonas Nick
    - Josh Rickmar
    - Olaoluwa Osuntokun
    - Oleg Andreev

Changes in 0.10.0 (Sun Mar 01 2015)
  - Protocol and network related changes:
    - Add a new checkpoint at block height 343185
    - Implement BIP066 which includes support for version 3 blocks, a new
      consensus rule which prevents non-DER encoded signatures, and a
      double-threshold switchover mechanism
    - Rather than announcing all known addresses on getaddr requests which
      can possibly result in multiple messages, randomize the results and
      limit them to the max allowed by a single message (1000 addresses)
    - Add more reserved IP spaces to the address manager
  - Transaction relay (memory pool) changes:
    - Make transactions which contain reserved opcodes nonstandard

doc.go
// Copyright (c) 2013-2016 The btcsuite developers
// Copyright (c) 2015-2017 The Decred developers
// Use of this source code is governed by an ISC
// license that can be found in the LICENSE file.

/*
pfcd is a full-node PicFight coin implementation written in Go.

The default options are sane for most users.  This means pfcd will work 'out of
the box' for most users.  However, there are also a wide variety of flags that
can be used to control it.

The following section provides a usage overview which enumerates the flags.  An
interesting point to note is that the long form of all of these options
(except -C) can be specified in a configuration file that is automatically
parsed when pfcd starts up.  By default, the configuration file is located at
~/.pfcd/pfcd.conf on POSIX-style operating systems and %LOCALAPPDATA%\pfcd\pfcd.conf
on Windows.  The -C (--configfile) flag, as shown below, can be used to override
this location.

Usage:
  pfcd [OPTIONS]

Application Options:
  -V, --version             Display version information and exit
  -C, --configfile=         Path to configuration file
  -b, --datadir=            Directory to store data
      --logdir=             Directory to log output.
      --nofilelogging=      Disable file logging.
  -a, --addpeer=            Add a peer to connect with at startup
      --connect=            Connect only to the specified peers at startup
      --nolisten            Disable listening for incoming connections -- NOTE:
                            Listening is automatically disabled if the --connect
                            or --proxy options are used without also specifying
                            listen interfaces via --listen
      --listen=             Add an interface/port to listen for connections
                            (default all interfaces port: 9108, testnet: 19108)
      --maxsameip=          Max number of connections with the same IP -- 0 to
                            disable (default: 5)
      --maxpeers=           Max number of inbound and outbound peers (125)
      --nobanning           Disable banning of misbehaving peers
      --banduration=        How long to ban misbehaving peers.  Valid time units
                            are {s, m, h}.  Minimum 1 second (24h0m0s)
      --banthreshold=       Maximum allowed ban score before disconnecting and
                            banning misbehaving peers.
      --whitelist=          Add an IP network or IP that will not be banned.
                            (eg. 192.168.1.0/24 or ::1)
  -u, --rpcuser=            Username for RPC connections
  -P, --rpcpass=            Password for RPC connections
      --rpclimituser=       Username for limited RPC connections
      --rpclimitpass=       Password for limited RPC connections
      --rpclisten=          Add an interface/port to listen for RPC connections
                            (default port: 9109, testnet: 19109)
      --rpccert=            File containing the certificate file
      --rpckey=             File containing the certificate key
      --rpcmaxclients=      Max number of RPC clients for standard connections
                            (10)
      --rpcmaxwebsockets=   Max number of RPC websocket connections (25)
      --norpc               Disable built-in RPC server -- NOTE: The RPC server
                            is disabled by default if no rpcuser/rpcpass or
                            rpclimituser/rpclimitpass is specified
      --notls               Disable TLS for the RPC server -- NOTE: This is only
                            allowed if the RPC server is bound to localhost
      --nodnsseed           Disable DNS seeding for peers
      --externalip=         Add an ip to the list of local addresses we claim to
                            listen on to peers
      --proxy=              Connect via SOCKS5 proxy (eg. 127.0.0.1:9050)
      --proxyuser=          Username for proxy server
      --proxypass=          Password for proxy server
      --onion=              Connect to tor hidden services via SOCKS5 proxy
                            (eg. 127.0.0.1:9050)
      --onionuser=          Username for onion proxy server
      --onionpass=          Password for onion proxy server
      --noonion             Disable connecting to tor hidden services
      --torisolation        Enable Tor stream isolation by randomizing user
                            credentials for each connection.
      --testnet             Use the test network
      --simnet              Use the simulation test network
      --regnet              Use the regression test network
      --nocheckpoints       Disable built-in checkpoints.  Don't do this unless
                            you know what you're doing.
      --dbtype=             Database backend to use for the Block Chain (ffldb)
      --profile=            Enable HTTP profiling on given [addr:]port -- NOTE: port
                            must be between 1024 and 65536
      --cpuprofile=         Write CPU profile to the specified file
      --memprofile=         Write mem profile to the specified file
      --dumpblockchain=     Write blockchain as a gob-encoded map to the
                            specified file
      --miningtimeoffset=   Offset the mining timestamp of a block by this many
                            seconds (positive values are in the past)
  -d, --debuglevel=         Logging level for all subsystems {trace, debug,
                            info, warn, error, critical} -- You may also specify
                            <subsystem>=<level>,<subsystem2>=<level>,... to set
                            the log level for individual subsystems -- Use show
                            to list available subsystems (info)
      --upnp                Use UPnP to map our listening port outside of NAT
      --minrelaytxfee=      The minimum transaction fee in PFC/kB to be
                            considered a non-zero fee.
      --limitfreerelay=     Limit relay of transactions with no transaction fee
                            to the given amount in thousands of bytes per
                            minute (15)
      --norelaypriority     Do not require free or low-fee transactions to have
                            high priority for relaying
      --maxorphantx=        Max number of orphan transactions to keep in memory
                            (1000)
      --generate            Generate (mine) bitcoins using the CPU
      --miningaddr=         Add the specified payment address to the list of
                            addresses to use for generated blocks -- At least
                            one address is required if the generate option is
                            set
      --blockminsize=       Mininum block size in bytes to be used when creating
                            a block
      --blockmaxsize=       Maximum block size in bytes to be used when creating
                            a block (375000)
      --blockprioritysize=  Size in bytes for high-priority/low-fee transactions
                            when creating a block (20000)
      --getworkkey=         DEPRECATED -- Use the --miningaddr option instead
      --nonaggressive       Disable mining off of the parent block of the blockchain
                            if there aren't enough voters
      --nominingstatesync   Disable synchronizing the mining state with other nodes
      --allowoldvotes       Enable the addition of very old votes to the mempool

      --sigcachemaxsize=    The maximum number of entries in the signature
                            verification cache.
      --blocksonly          Do not accept transactions from remote peers.
      --acceptnonstd        Accept and relay non-standard transactions to
                            the network regardless of the default settings
                            for the active network.
      --rejectnonstd        Reject non-standard transactions regardless of the
                            default settings for the active network.
      --altdnsnames:        Specify additional dns names to use when
                            generating the rpc server certificate
                            [supports PFCD_ALT_DNSNAMES environment variable]

Help Options:
  -h, --help           Show this help message

*/
package main
